### ISSUE TITLE
Donate Metadata Editor to Lab

### ISSUE THREAD

**Issue by marthacryan** [elyra-ai/elyra#1997]
This is a planning / discussion issue for donating the metadata editor to JupyterLab. 

**Context**
- Use it to implement new Settings UI (https://github.com/jupyterlab/jupyterlab/issues/8148)
- Design already done, but needs implementation
- Potential connection: thinking about leveraging the metadata editor / contributing that code
- Existing settings editor: https://github.com/jupyterlab/jupyterlab/blob/master/packages/settingeditor/src/settingeditor.tsx
- Existing (currently not maintained) metadata service extension: https://github.com/jupyterlab/jupyterlab-metadata-service

**Phases of Donation:**

Backend:
- Move metadata services into jupyter server

Frontend:

1. Move metadata editor "as is" into core (can be done in parallel with backend move)
2. Updated metadata editor to work with donated backend
3. Update settings editor to use metadata editor instead of json editor
4. Update settings editor to look and work more like wireframe
5. Update Elyra to leverage donated metadata code

**Comment by kevin-bates** [comment]
Hmm - this is a bit of a surprise since it sounds like a ton of UI work was done by the UCI team and Alex had even [volunteered to help get it into core](https://github.com/jupyterlab/jupyterlab/issues/8148#issuecomment-683086898).  Are we duplicating effort and/or stepping on toes?

Regarding the backend, I believe we need to support [Bring-your-own-schemas](https://github.com/elyra-ai/elyra/discussions/1672) and promote Namespaces (and BYO Namespaces) prior to any move to core.  Currently, namespaces are just string-valued attributes on schemas and schemas are static.

**Comment by ajbozarth** [comment]
> Hmm - this is a bit of a surprise since it sounds like a ton of UI work was done by the UCI team and Alex had even volunteered to help get it into core. Are we duplicating effort and/or stepping on toes?

The work done so far was all design, no implementation, and based on their design our thinking was that since we were already planning on donating the Metadata Editor to core then we should leverage it for the implementation of the setting UI which we also volunteered to work on.

As for the donation in general, we did discuss the idea that if the Setting UI work is ready before the backend metadata service is donated then we can submit only the metadata code that is leveraged by the settings ui and leave the remaining UI code tied to the service as a draft until the backend is committed

**Comment by kevin-bates** [comment]
Just stumbled upon this comment which essentially reiterates the tasks I mentioned previously relative to the server effort: https://github.com/elyra-ai/elyra/issues/701#issuecomment-716888388

If the server portion was implemented as a server extension, does that preclude the metadata editor from being placed in core?

**Comment by ajbozarth** [comment]
> If the server portion was implemented as a server extension, does that preclude the metadata editor from being placed in core?

Yes and no, the metadata editor can be partially donated without the backend if we leave any code related to calling the service in Elyra and only move the actual editor (this is essentially the case I described above if the setting UI is ready before the server extension). Also the "server extension" would need to be included in core jupyterlab, either as part of jupyterserver or in jupyterlab itself, if we leave it as a "3rd party" extension then we would not be able to put code in core that relies on it.

**Comment by kevin-bates** [comment]
Ok.  So, without the server piece in the backend core, you'd essentially be donating a general editor that drives itself from JSON schema?  This could be used for settings and even the soon-to-be-needed kernel parameter editor that captures parameters prior to a kernel's startup. I.e., instances whose retrieval and persistence are not reliant on the "metadata service".  But using that same editor to persist Elyra metadata instances (where retrieval/persistence occurs via the "metadata service" - baked into the server or as an extension) could not occur - correct?

**Comment by ajbozarth** [comment]
IIUC that is accurate

**Comment by marthacryan** [comment]
Due to changes on the JupyterLab side from https://github.com/jupyterlab/jupyterlab/pull/11079, we won't be donating the metadata editor anymore - we're updating the metadata editor on our side to reuse these changes in https://github.com/elyra-ai/elyra/pull/2464.
